core/src/main/scala/org/apache/spark/storage/ShuffleBlockFetcherIterator.scala

```diff
@@ -379,6 +452,13 @@ final class ShuffleBlockFetcherIterator(
     // Get Local Blocks
     fetchLocalBlocks()
     logDebug(s"Got local blocks in ${Utils.getUsedTimeNs(startTimeNs)}")
+
+    // Get Host-local Blocks
+    if (hostLocalBlocks.nonEmpty) {
+      val hostLocalStartTimeNs = System.nanoTime()
+      fetchHostLocalBlocks()
```

Comment:
I think there is a corner case that is missed here.  What if there are *only* host-local blocks, but none of them have any hostLocalDirs available?  You'd put those remote requests in the fetch queue, but you'd never send them out until a call to `fetchUpToMaxBytes()` inside `next()` -- but you'd never get there because `next()` would block forever waiting to dequeue.

I think you can fix this just by adding another call to `fetchUpToMaxBytes()` here.  Definitely worth a comment explaining this as well.

Reply:
It is a good catch. Thanks! Fixed and a unit test is added for this case.